### PR TITLE
feat: implementation of search box expandable

### PR DESCRIPTION
**Pull request by mfazil-eightfold**
## SUMMARY:
Added expandable search box

https://user-images.githubusercontent.com/109717425/193663166-b05abebf-f632-4686-8b23-63be8146ca16.mov

## JIRA TASK (Eightfold Employees Only):
https://eightfoldai.atlassian.net/browse/ENG-30444

## CHANGE TYPE:

-   [ ] Bugfix Pull Request
-   [x] Feature Pull Request

## TEST COVERAGE:

-   [ ] Tests for this change already exist
-   [ ] I have added unittests for this change

### CODE REVIEW

File: src/components/Inputs/Input.types.ts [hunk 1]
```diff
@@ -302,4 +302,9 @@ export interface InputProps<T>
      * @default 10
      */
     waitInterval?: number;
+    /**
+     * Make input expandable on focus
+     * @default false
+     */
+    expandable?: boolean;
```
Comment (jhoward-eightfold) [inline src/components/Inputs/Input.types.ts:309]: alpha sort props

File: src/components/Inputs/TextInput/TextInput.tsx [hunk 2]
```diff
@@ -70,6 +70,7 @@ export const TextInput: FC<TextInputProps> = React.forwardRef(
             style,
             value,
             waitInterval = 10,
+            expandable = false,
```
Comment (jhoward-eightfold) [inline src/components/Inputs/TextInput/TextInput.tsx:73]: alpha sort

File: src/components/Inputs/TextInput/TextInput.tsx [hunk 3]
```diff
@@ -392,88 +394,120 @@ export const TextInput: FC<TextInputProps> = React.forwardRef(
...
+                            placeholder={placeholder}
+                            required={required}
+                            role="textbox"
+                            style={style}
+                            tabIndex={0}
+                            type={numbersOnly ? 'number' : htmlType}
+                            value={inputValue}
+                            readOnly={readonly}
+                        />
+                        {expandable && iconButtonProps && (
```
Comment (jhoward-eightfold) [inline src/components/Inputs/TextInput/TextInput.tsx:424]: what happens when expandable and no iconButtonProps?

File: src/components/Inputs/TextInput/TextInput.tsx [hunk 4]
```diff
@@ -392,88 +394,120 @@ export const TextInput: FC<TextInputProps> = React.forwardRef(
...
+                                        mergedSize
+                                    )}
+                                    htmlType={iconButtonProps.htmlType}
                                 />
                             )}
+                            {clearable &&
+                                clearButtonShown &&
+                                !numbersOnly &&
+                                htmlType !== 'number' && (
+                                    <DefaultButton
```
Comment (jhoward-eightfold) [inline src/components/Inputs/TextInput/TextInput.tsx:494]: was this clearable button always outside of the input? It seems a bit oddly placed

File: src/components/Inputs/input.module.scss [hunk 5]
```diff
@@ -1438,6 +1447,93 @@
...
+                input.input-#{$size-name}:placeholder-shown {
+                min-width: $size-value;
+            }
+
+            .input-group:not(:hover)
+                input.input-#{$size-name}:placeholder-shown {
+                padding: 0;
+                overflow: hidden;
+
+                &::placeholder {
```
Comment (jhoward-eightfold) [inline src/components/Inputs/input.module.scss:1527]: This is unlikely to work in ie11... does it get polyfilled in compilation?
https://caniuse.com/?search=placeholder

File: src/components/Inputs/input.module.scss [hunk 6]
```diff
@@ -1438,6 +1447,93 @@
...
+                    font-size: 0;
+                }
+
+                ~ .actionWrapper {
```
Comment (jhoward-eightfold) [inline src/components/Inputs/input.module.scss:1531]: can you be more precise with your selector?

File: src/components/Inputs/input.module.scss [hunk 7]
```diff
@@ -1438,6 +1447,93 @@
...
+        }
+
+        // Unset Rules
+        @each $size-name, $size-value in $input-size-map {
+            &.input-#{$size-name} > .expandable-wrapper {
+                min-width: $size-value;
+                min-height: $size-value;
+            }
+
+            .input-group:not(:hover)
```
Comment (jhoward-eightfold) [inline src/components/Inputs/input.module.scss:1517]: seemingly focus or focus-within might also be relevant here?
strange selector wrapping and indentation

File: src/styles/themes/_definitions.scss [hunk 8]
```diff
@@ -88,6 +88,18 @@ $eightfold-logo-gradient-dark: linear-gradient(
         #1abbcd 50.02%,
         #6fc6aa 100.03%
     );
+$expandable-header-gradient: linear-gradient(
+    270deg,
+    hsla(0, 0%, 100%, 0) 22.25%,
+    hsla(0, 0%, 100%, 0.995) 41.17%,
+    #fff 80.83%
```
Comment (jhoward-eightfold) [inline src/styles/themes/_definitions.scss:95]: this needs to be theme var driven

File: src/styles/themes/_definitions.scss [hunk 9]
```diff
@@ -88,6 +88,18 @@ $eightfold-logo-gradient-dark: linear-gradient(
...
+);
+// $expandable-header-gradient-reverse: linear-gradient(
+//     -90deg,
+//     rgba(254, 254, 254, 0) 22.25%,
+//     rgba(255, 255, 255, 0.994792) 41.17%,
+//     #ffffff 80.83%
+// );
```
Comment (jhoward-eightfold) [inline src/styles/themes/_definitions.scss:102]: cleanup?

**Review by jhoward-eightfold — CHANGES_REQUESTED** [review]

File: src/components/Inputs/input.module.scss (hunk 5 above)
Comment (jhoward-eightfold) [reply src/components/Inputs/input.module.scss:1527]: Additionally, the instant change to font-size seems to cause an unsightly overlap of text and search icon during the animation. You might want to rethink this approach.

Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1527]: will add a polyfill for this.

File: src/components/Inputs/TextInput/TextInput.tsx (hunk 4 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/TextInput/TextInput.tsx:494]: yes, i did not change the order.

File: src/components/Inputs/TextInput/TextInput.tsx (hunk 3 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/TextInput/TextInput.tsx:424]: The idea/assumption is that the user will pass `iconButtonProps`.
There will be not placeholder in case `iconButtonProps` is not provided.

File: src/components/Inputs/input.module.scss (hunk 7 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1517]: some last moment change. will fix this.

File: src/styles/themes/_definitions.scss (hunk 9 above)
Comment (mfazil-eightfold) [reply src/styles/themes/_definitions.scss:102]: some last moment change. will fix this.

Comment (mfazil-eightfold) [reply src/styles/themes/_definitions.scss:102]: will remove this.

File: src/components/Inputs/input.module.scss (hunk 6 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1531]: This is required because we want to select a sibling with the class name action-wrapper. BTW, I've renamed actionWrapper to action-wrapper

File: src/components/Inputs/input.module.scss (hunk 7 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1517]: We're setting the input style when the input-group has not hovered.

File: src/styles/themes/_definitions.scss (hunk 8 above)
Comment (mfazil-eightfold) [reply src/styles/themes/_definitions.scss:95]: these are directly from design and we don't have such colors.

File: src/components/Inputs/Input.types.ts [hunk 10]
```diff
@@ -209,6 +209,17 @@ export interface InputProps<T>
...
+     * Make input expandable on focus.
+     * Default slides from left upon setting this to true.
+     * @default false
+     */
+    expandable?: boolean;
+    /**
+     * Make input expandable slide from right.
+     * @default false
+     */
+    expandRight?: boolean;
```
Comment (jhoward-eightfold) [inline src/components/Inputs/Input.types.ts:222]: @dkilgore-eightfold thoughts on this naming from a RTL perspective?

File: src/components/Inputs/input.module.scss (hunk 7 above)
Comment (jhoward-eightfold) [reply src/components/Inputs/input.module.scss:1532]: I understand the selector, but is this sufficient? When a user tabs into the input, things don't look as they should. Not sure if this is directly related, but needs to be addressed.

File: src/styles/themes/_definitions.scss (hunk 8 above)
Comment (jhoward-eightfold) [reply src/styles/themes/_definitions.scss:95]: we do have a #fff color var (its background or something). We should use it. If someone implements a dark color theme for instance, we don't want this remaining white.

File: src/components/Inputs/input.module.scss [hunk 11]
```diff
@@ -1438,6 +1447,114 @@
...
+                    font-size: 0;
+                }
+
+                // Edge
+                &::-ms-input-placeholder {
+                    font-size: 0;
+                }
+
+                &::placeholder {
+                    font-size: 0;
```
Comment (jhoward-eightfold) [inline src/components/Inputs/input.module.scss:1549]: font-size based change here is still resulting in overlap of the search text and icons during the animation/expansion.

Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1567]: Increased the transition speed. It's bearly visible during expansion.

File: src/styles/themes/_definitions.scss [hunk 12]
```diff
@@ -88,6 +88,12 @@ $eightfold-logo-gradient-dark: linear-gradient(
         #1abbcd 50.02%,
         #6fc6aa 100.03%
     );
+$expandable-header-gradient: linear-gradient(
```
Comment (dkilgore-eightfold) [inline src/styles/themes/_definitions.scss:91]: rtl version?

File: src/components/Inputs/input.module.scss [hunk 13]
```diff
@@ -1438,6 +1447,132 @@
     &.input-stretch {
         width: 100%;
     }
+
+    &.isExpandable {
```
Comment (dkilgore-eightfold) [inline src/components/Inputs/input.module.scss:1451]: RTL is also required.

File: src/components/Inputs/input.module.scss [hunk 14]
```diff
@@ -1438,6 +1447,132 @@
...
+            left: -3px;
+        }
+
+        .overlay {
+            position: absolute;
+            left: 0;
+            top: 0;
+            height: 100%;
+            width: 200%;
+            z-index: -1;
```
Comment (dkilgore-eightfold) [inline src/components/Inputs/input.module.scss:1481]: why is this needed?

File: src/components/Inputs/input.module.scss (hunk 11 above)
Comment (dkilgore-eightfold) [reply src/components/Inputs/input.module.scss:1567]: It's still noticeable. You need to add a mask.

File: src/components/Inputs/Input.types.ts (hunk 10 above)
Comment (dkilgore-eightfold) [reply src/components/Inputs/Input.types.ts:222]: This is fine, but this feature still does not yet support canvas direction. Added comments in CSS. This prop will not work for loc.

**Review by dkilgore-eightfold — CHANGES_REQUESTED** [review]

File: src/components/Inputs/input.module.scss [hunk 15]
```diff
@@ -1438,6 +1447,132 @@
     &.input-stretch {
         width: 100%;
     }
+
+    &.isExpandable {
+        .expandable-wrapper {
+            position: relative;
+        }
+
+        .input-group {
```
Comment (dkilgore-eightfold) [inline src/components/Inputs/input.module.scss:1456]: inputs are highly configurable. You'll need to test your animation against a few cases:

- Search icon button only
- Icon + Search Icon Button
- No Icon or Icon button
- Align Icon (left and right)
- RTL canvas dir when align icon is left

Comment (dkilgore-eightfold) [reply src/components/Inputs/input.module.scss:1456]: Additionally, you'll need to test underline and rectangle shaped input all the above.

Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1456]: this is done

File: src/components/Inputs/input.module.scss (hunk 14 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1476]: Need to show a gradient background.

File: src/components/Inputs/input.module.scss [hunk 16]
```diff
@@ -1505,3 +1637,7 @@
 }
 
 @import './Styles/rtl';
+
+// *:focus {
```
Comment (dkilgore-eightfold) [inline src/components/Inputs/input.module.scss:1641]: remove

File: src/components/Inputs/input.module.scss [hunk 17]
```diff
@@ -1438,6 +1447,129 @@
...
+            }
+
+            .input-group:not(:hover):not(:focus-within)
+                input.input-#{$size-name}:placeholder-shown {
+                min-width: $size-value;
+                padding: 0;
+                overflow: hidden;
+
+                // Internet Explorer 10–11
+                &:-ms-input-placeholder {
```
Comment (dkilgore-eightfold) [inline src/components/Inputs/input.module.scss:1554]: Rather than `font-size`, you might try `opacity` and `width`, then animate them to prevent the overlap when the `input-group `meets the criteria to expand.

**Review by dkilgore-eightfold — CHANGES_REQUESTED** [review]

File: src/components/Inputs/input.module.scss (hunk 17 above)
Comment (mfazil-eightfold) [reply src/components/Inputs/input.module.scss:1554]: ahh! I should have done that only. thanks for the trick.

Comment (dkilgore-eightfold) [reply src/components/Inputs/input.module.scss:1554]: `width` may not even be needed as `::placeholder` is a pseudo selector.

**Review by dkilgore-eightfold — APPROVED** [review]
